Fix bad interaction between SIL function serialization and weak linking

slavapestov opened the pull request:
If a declaration in a module is `@available(fooOS 20, *)`, it should still be weak linked when referenced from another module with a deployment target of fooOS 10, but not when referenced from any module built with a deployment target of fooOS 20.

Previously we would only serialize a single isWeakLinked bit in the SILFunction, which was computed by checking the availability of the original AST function declaration against the module's own deployment target, but this is insufficient if a SIL function is serialized and deserialized inside another module.

Instead, record the original OS version in which the declaration was introduced, and serialize that together with the SIL function.

Fixes <rdar://problem/52783668>.

slavapestov committed:
SIL: Micro-optimize SILFunction layout

slavapestov committed:
AST: Refactor Decl::isWeakImported() API a little bit

First, remove the AvailabilityContext parameter; it was confusing because
we actually always want to use the deployment target here.

Then, split this method up into three methods:

- isAlwaysWeakImported(): simply checks for a @_weakLinked attribute, either
  on the declaration itself or one of its parent contexts.

- getAvailabilityForLinkage(): returns the OS version availability when
  this declaration was introduced, or if the declaration does not have
  explicit availability, check it's storage (if its an accessor), or its
  parent contexts.

- isWeakImported(ModuleDecl *fromModule): combines these two checks to
  determine if the declaration should be weak linked when referenced from
  the given module, or if it might be weak referenced from some module
  (if the module parameter is null).

slavapestov commented:
I forgot to mark this as "WIP". Still needs tests

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

jrose-apple commented on include/swift/AST/Decl.h:
```diff
@@ -940,9 +940,27 @@ class alignas(1 << DeclAlignInBits) Decl {
-                      AvailabilityContext fromContext) const;
+  AvailabilityContext getAvailabilityForLinkage() const;
+
+  /// Whether this declaration or one of it's outer contexts has the
```
Typo: "its"

jrose-apple commented on include/swift/AST/Decl.h:
```diff
@@ -940,9 +940,27 @@ class alignas(1 << DeclAlignInBits) Decl {
+
+  /// Whether this declaration or one of it's outer contexts has the
+  /// @_weakLinked attribute.
+  bool isWeakImported() const;
```
Can you rename this so it's more obvious what it does? Basically no one should call it directly. (Maybe it should be a helper function in Decl.cpp instead.)

jrose-apple replied:
Sure, or `isUnconditionallyWeakImported()` maybe.

jrose-apple commented on include/swift/AST/Decl.h:
```diff
@@ -940,9 +940,27 @@ class alignas(1 << DeclAlignInBits) Decl {
+  /// If \p fromModule is non-null, always returns false if the
+  /// declaration is part of \p fromModule, or if the declaration has
+  /// availability but it is at least as available as the deployment
+  /// target.
```
This "deployment target" part is questionable, since it isn't actually related to `fromModule`. In practice, the only module that will be passed here is the one being compiled, but if you *did* pass another one it could give a technically-incorrect answer.

slavapestov replied:
I clarified the comment to explain that fromModule should either be null or the main module. I agree this is not an ideal API either, but it still seems better than alternatives I could think of.

jrose-apple commented on lib/SIL/SILFunctionBuilder.cpp:
```diff
@@ -162,8 +162,17 @@ SILFunctionBuilder::getOrCreateFunction(SILLocation loc, SILDeclRef constant,
+        auto fromContext = AvailabilityContext::forDeploymentTarget(
+          decl->getASTContext());
+        if (!fromContext.isContainedIn(containingContext))
+          F->setWeakLinked();
```
What's the difference between this whole block and `isWeakImported(/*fromModule*/nullptr)`? The fact that it's not obvious to me makes me concerned about misuse.

slavapestov replied:
Take a look at the next patch, this gets cleaned up again. Its just there to keep tests passing in the half-refactored state

jrose-apple replied:
Oh, GitHub reordered them again…

jrose-apple reviewed (commented):
Comments about `isWeakImported`

jrose-apple commented on include/swift/SIL/SILFunction.h:
```diff
@@ -213,8 +218,9 @@ class SILFunction
-  /// Whether cross-module references to this function should use weak linking.
-  unsigned IsWeakLinked : 1;
+  /// Whether cross-module references to this function should always use
+  ///weak linking.
```
Typo: missing space

jrose-apple commented on lib/AST/Decl.cpp:
```diff
@@ -631,10 +631,10 @@ AvailabilityContext Decl::getAvailabilityForLinkage() const {
   auto *dc = getDeclContext();
-  if (auto *nominal = dyn_cast<NominalTypeDecl>(dc))
-    return nominal->getAvailabilityForLinkage();
-  else if (auto *ext = dyn_cast<NominalTypeDecl>(dc))
```
Heh, oops.

jrose-apple commented on lib/SIL/SILFunction.cpp:
```diff
@@ -259,6 +262,29 @@ bool SILFunction::isTypeABIAccessible(SILType type) const {
+      fromModule->getASTContext());
+  return !fromContext.isContainedIn(Availability);
+
+  return false;
```
Unreachable return.

jrose-apple reviewed (commented):
Serialization stuff is actually pretty straightforward after that, I guess.

slavapestov committed:
SIL: Serialize availability with the SIL function, not just a weak imported flag

The weak imported flag is now only set if the attribute is unconditionally
weak linked, which is the case when it or one of its parent contexts has a
@_weakLinked attribute.

To correctly handle weak linking based availability with serialized SIL
functions, we need to serialize the actual version tuple when the SIL function
was introduced. This is because the deployment target of the client app can
be older than the deployment target that the original module was built with.

Fixes <rdar://problem/52783668>.

slavapestov commented:
@swift-ci Please smoke test

varungandhi-apple commented:
I don't feel sufficiently qualified to review the PR but I do have a high-level question: in some places we're using the terminology "weak linking" and in some places it is "weak import" -- I understand the former but not the latter. What does "weak import" mean/how does it differ from weak linking? AFAICT, there is not attribute @_weak or something on an import clause here.

slavapestov commented:
@varungandhi-apple It was like that before my PR too. I'm not quite sure to be honest, it might just be an accident of history. I did rename SILFunction::isWeakLinked() to isWeakImported() to be consistent with the method on Decl.

slavapestov commented:
@swift-ci Please smoke test macOS
